brain_friendly.py

```diff
@@ -32,31 +32,20 @@ def eval_program(program, index, array, input=None, output=None):
                 # This sends 128 -> -128, and 255 -> -1
                 if array[index] > 127:
                     array[index] -= 256
+
         elif command == "[":
             # "[" jumps past the matching "[" if the current cell is 0.
-
-            # Pushing this is fine even if the current cell is 0
-            # Because it will be poped when we jump to the "]"
-            bracket_stack.append(program_index)
-
-            # Search for next "]"
             if array[index] == 0:
-                bracket_count = 0
-                while bracket_count != -1 :
-                    program_index += 1
-                    if program[program_index] == "[":
-                        bracket_count += 1
-                    elif program[program_index] == "]":
-                        bracket_count -= 1
-                    
+                    program_index = brace_dict[program_index]
+
         elif command == "]":
             # "]" jumps back to the matching "[" if the current cell is NOT 0
             if array[index] != 0:
                 #Go to matching "["
-                program_index = bracket_stack[-1]
-            else:
-                # Fall out of loop, pop matching "["
-                bracket_stack.pop()
+                for left, right in brace_dict.iteritems():
```

Comment:
Instead of doing this terrible linear search, store both `left: right` and `right: left` in the dictionary, so you can do `program_index = brace_dict[program_index]` for both directions.